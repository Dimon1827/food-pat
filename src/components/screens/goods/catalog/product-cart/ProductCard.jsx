import './ProductCard.scss';
import ProductCardInformation from './ProductCardInformation';
import TaimerList from './taimer/TaimerList';

function ProductCard({ name, ingredients, image, price, oldPrice, time }) {
  return (
    <div
      className={`product-card ${
        !!oldPrice ? 'additional-products__card' : 'catalog__item'
      }`}
    >
      <a href="404.html" className="product-card__link">
        <picture>
          <source
            type="image/webp"
            srcSet={`${image}.webp`}
            width="280"
            height="240"
          />
          <img
            className={`product-card__image ${
              !!oldPrice ? 'additional-products__image' : ''
            }`}
            src={`${image}.${!!oldPrice ? 'png' : 'jpg'}`}
            alt={name}
            width="280"
            height="240"
          />
        </picture>
      </a>
      {!!oldPrice && <TaimerList time={time} />}
      <ProductCardInformation
        name={name}
        ingredients={ingredients}
        price={price}
        oldPrice={oldPrice}
      />
    </div>
  );
}

export default ProductCard;

// additional-products__list надо с ним что то делать
