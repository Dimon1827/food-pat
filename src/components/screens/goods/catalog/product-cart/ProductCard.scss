@import '@/assets/styles/variables';

.product-card {
  border-radius: 18px;
  box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.08);
  background: $white-color;
  max-width: 417px;
  margin: 0 auto;

  &:not(:last-child) {
    margin-bottom: 20px;
  }

  @media (min-width: $tablet-width) {
    &:not(:last-child) {
      margin: 0;
    }
  }
}

.product-card__wrapper {
  padding: 15px;
}

.product-card__image {
  margin-bottom: 20px;
  border-radius: 20px;
  width: 100%;

  @media (min-width: $tablet-width) {
    margin-bottom: 30px;
  }
}

.product-card__text {
  font-weight: 700;
  font-size: 20px;
  line-height: 120%;
  margin: 0;
  margin-bottom: 15px;
  display: inline-block;

}

.product-card__ingredients {
  margin: 0;
  margin-bottom: 15px;
  font-style: italic;
}


.product-card__top-wrapper {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 20px;

  @media (min-width: $tablet-width) {
    margin-bottom: 26px;
  }
}

.product-card__link {
  text-decoration: none;

  &:first-child {
    justify-self: center;
    align-self: center;
    grid-row: 1/3;

    @media (min-width: $tablet-width) {
      grid-row: 1/2;
    }
  }
}

.product-card__subtitle {
  width: 100px;
  font-size: 21px;
  color: #444444;
  line-height: 145%;

  @media (min-width: $tablet-width) {
    width: 135px;
  }
}

.product-card__button {
  padding: 10px 24px;
  background-color: #f03328;
  border-radius: 38px;
  min-width: 100px;
  max-width: 148px;
}
