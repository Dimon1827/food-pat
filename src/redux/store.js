import { configureStore, combineReducers } from '@reduxjs/toolkit';
import {
  persistReducer,
  persistStore,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import productsReducer from './slice/productsSlice';

const persistConfig = {
  key: 'root',
  version: 2,
  storage,
};

const reducer = combineReducers({
  products: productsReducer,
});

const persistedReducer = persistReducer(persistConfig, reducer);

const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }),
});

export const persistor = persistStore(store);

export default store;
