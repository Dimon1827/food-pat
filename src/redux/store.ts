import { configureStore, combineReducers } from '@reduxjs/toolkit';
import { useDispatch } from 'react-redux';
import {
  persistReducer,
  persistStore,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import productsReducer from './slice/productsSlice';

const persistConfig = {
  key: 'root',
  version: 2,
  storage,
};

const reducer = combineReducers({
  products: productsReducer,
});

const persistedReducer = persistReducer(persistConfig, reducer);

const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }),
});

export type RootState = ReturnType<typeof persistedReducer>;
export type AppDispatch = typeof store.dispatch;
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const persistor = persistStore(store);
export default store;
