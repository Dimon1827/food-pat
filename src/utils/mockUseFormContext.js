import { useFormContext } from "react-hook-form";

export function mockUseFormContext(errors) {
  useFormContext.mockReturnValue({
    register: jest.fn(),
    formState: {
      errors: errors,
    },
  });
}


