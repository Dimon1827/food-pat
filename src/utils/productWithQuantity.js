const productWithQuantity = (food) => {
  return {
    ...food,
    isChecked: false,
    quantity: 1,
  };
};

export default productWithQuantity;
