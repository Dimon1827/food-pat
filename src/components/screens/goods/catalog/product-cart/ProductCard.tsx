import { Link } from 'react-router-dom';
import { IProductWithDiscounts } from '@/types';
import ProductCardInformation from './ProductCardInformation';
import TaimerList from './taimer/TaimerList';
import './ProductCard.scss';

function ProductCard({
  name,
  ingredients,
  image,
  price,
  oldPrice,
  time,
  id,
}: IProductWithDiscounts) {
  return (
    <div
      className={`product-card ${
        !!oldPrice ? 'additional-products__card' : 'catalog__item'
      }`}
    >
      <Link to={name}>
        <picture>
          <source
            type="image/webp"
            srcSet={`${image}.webp`}
            width="280"
            height="240"
          />
          <img
            className={`product-card__image ${
              !!oldPrice ? 'additional-products__image' : ''
            }`}
            src={`${image}.${!!oldPrice ? 'png' : 'jpg'}`}
            alt={name}
            width="280"
            height="240"
          />
        </picture>
      </Link>
      {!!oldPrice && <TaimerList time={time} />}
      <ProductCardInformation
        name={name}
        ingredients={ingredients}
        price={price}
        oldPrice={oldPrice}
        id={id}
        image={image}
      />
    </div>
  );
}

export default ProductCard;