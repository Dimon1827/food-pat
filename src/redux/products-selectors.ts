import { RootState } from "./store";

export const selectProducts = (state: RootState) => state.products.products;
export const selectTotalPrice = (state: RootState) => state.products.totalPrice;
export const selectTotalQuantity = (state: RootState) => state.products.totalQuantity;
