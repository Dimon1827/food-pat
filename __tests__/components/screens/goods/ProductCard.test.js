import ProductCard from '@/components/screens/goods/catalog/product-cart/ProductCard';
import { renderWithRouter } from '@/utils/renderWithRouter';
import * as store from '@/redux/store';
import { screen } from '@testing-library/react';

const ProductCardInformation = {
  id: '123',
  name: 'Салат с фруктами',
  ingredients: 'Апельсин, зелень, огурцы, банан',
  description:
    'Свежий и витаминный десерт, который зарядит вас энергией! Сочетание разноцветных фруктов - клубника, киви, банан, яблоко, груша - заправленный нежным йогуртом или медом. Идеально для легкого перекуса или десерта',
  oldPrice: 500,
  price: 400,
  time: {
    days: 3,
    hours: 11,
    minutes: 18,
  },
  image: './assets/images/day-item-1',
};

const mockedDispatch = jest.spyOn(store, 'useAppDispatch');

describe('ProductCard', () => {
  it('ProductCard renders correctly', () => {
    mockedDispatch.mockReturnValue(jest.fn());
    renderWithRouter(
      <ProductCard {...ProductCardInformation} />
    );
    expect(screen.getByTestId('product-card')).toMatchSnapshot();
  });
  it('ProductCard shouldnt have render taimer component without old price and should have another classes', () => {
    mockedDispatch.mockReturnValue(jest.fn());
    renderWithRouter(
      <ProductCard {...ProductCardInformation} oldPrice={null} />
    );
    expect(screen.getByTestId('product-card')).not.toHaveClass(
      'additional-products__card'
    );
    expect(screen.getByRole('img')).not.toHaveClass(
      'additional-products__image'
    );
    expect(screen.queryByTestId('taimer')).not.toBeInTheDocument();
  });
});
