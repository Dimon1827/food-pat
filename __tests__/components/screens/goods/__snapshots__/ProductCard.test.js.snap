// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard ProductCard renders correctly 1`] = `
<div
  class="product-card additional-products__card"
  data-testid="product-card"
>
  <a
    href="/Салат с фруктами"
  >
    <picture>
      <source
        height="240"
        srcset="./assets/images/day-item-1.webp"
        type="image/webp"
        width="280"
      />
      <img
        alt="Салат с фруктами"
        class="product-card__image additional-products__image"
        height="240"
        src="./assets/images/day-item-1.png"
        width="280"
      />
    </picture>
  </a>
  <ul
    class="additional-products__list"
  >
    <li
      class="additional-products__item"
    >
      <span
        class="additional-products__number"
      >
        3
      </span>
      <p
        class="additional-products__text"
      >
        дня
      </p>
    </li>
    <li
      class="additional-products__item"
    >
      <span
        class="additional-products__number"
      >
        11
      </span>
      <p
        class="additional-products__text"
      >
        часов
      </p>
    </li>
    <li
      class="additional-products__item"
    >
      <span
        class="additional-products__number"
      >
        18
      </span>
      <p
        class="additional-products__text"
      >
        минут
      </p>
    </li>
  </ul>
  <div
    class="product-card__wrapper"
    data-testid="product-card-information"
  >
    <div
      class="product-card__top-wrapper"
    >
      <a
        class="product-card__link"
        href="/Салат с фруктами"
      >
        <h2
          class="product-card__subtitle subtitle"
        >
          Салат с фруктами
        </h2>
      </a>
      <button
        class="product-card__button button"
      >
        Купить
      </button>
    </div>
    <p
      class="product-card__text"
    >
      Ингредиенты:
    </p>
    <p
      class="product-card__ingredients"
    >
      Апельсин, зелень, огурцы, банан
    </p>
    <p
      class="product-card__text"
    >
      Цена:
    </p>
    <span
      class="food__definition additional-products__old-price"
    >
      500
       ₽
    </span>
    <span
      class="food__definition food__price"
    >
      400
       ₽
    </span>
  </div>
</div>
`;
